### PR TITLE
Bluetooth: Controller: Updated ISO-AL time wrapping and restructured unit test

### PR DESCRIPTION
**Pull request by niag-Oticon**
-- Handle time calculations across the point at which the controller's time information from the controller's clock will wrap (in the
   microsecond range)
-- Use wrapped time in all ISO-AL time calculations
-- Included TX unit tests
-- Re-factored unit tests to use FFF

### CODE REVIEW

File: subsys/bluetooth/controller/ll_sw/isoal.c [hunk 1]
```diff
@@ -39,6 +42,14 @@
 					   PDU_ISO_SEG_TIMEOFFSET_SIZE)
 #endif /* CONFIG_BT_CTLR_ADV_ISO || CONFIG_BT_CTLR_CONN_ISO */
 
+/* Defined the wrapping point and mid point in the range of time input values,
+ * which depend on range of the controller's clock in microseconds.
+ */
+#define ISOAL_TIME_WRAPPING_POINT         (HAL_TICKER_TICKS_TO_US(HAL_TICKER_CNTR_MASK))
+#define ISOAL_TIME_MID_POINT              (ISOAL_TIME_WRAPPING_POINT / 2)
+#define ISOAL_TIME_SPAN_FULL              (ISOAL_TIME_WRAPPING_POINT + 1)
+#define ISOAL_TIME_SPAN_HALF              (ISOAL_TIME_SPAN_FULL / 2)
```
Comment (Thalley) [inline subsys/bluetooth/controller/ll_sw/isoal.c:51]: Why is `SPAN_HALF` not `ISOAL_TIME_SPAN_FULL / 2`?

File: subsys/bluetooth/controller/ll_sw/isoal.c [hunk 2]
```diff
@@ -89,6 +100,22 @@ isoal_status_t isoal_reset(void)
 	return err;
 }
 
+/**
+ * @breif Wraps given time within the range of 0 to ISOAL_TIME_WRAPPING_POINT
```
Comment (Thalley) [inline subsys/bluetooth/controller/ll_sw/isoal.c:104]: `@brief`

File: subsys/bluetooth/controller/ll_sw/isoal.c (hunk 1 above)
Comment (Thalley) [inline subsys/bluetooth/controller/ll_sw/isoal.c:51]: IMO it's typically useful to denote the time unit in macros like these (s, ms, us)

File: subsys/bluetooth/controller/ll_sw/isoal.c [hunk 3]
```diff
@@ -89,6 +100,22 @@ isoal_status_t isoal_reset(void)
...
+ */
+static uint32_t isoal_get_wrapped_time(uint32_t time_now, int32_t time_diff)
+{
+	LL_ASSERT(time_now <= ISOAL_TIME_WRAPPING_POINT);
+
+	uint32_t result = ((uint64_t)time_now + ISOAL_TIME_SPAN_FULL + time_diff) %
+				((uint64_t)ISOAL_TIME_SPAN_FULL);
+
+	return result;
+}
```
Comment (Thalley) [inline subsys/bluetooth/controller/ll_sw/isoal.c:117]: Ditto, it's IMO useful to denote the time units

File: subsys/bluetooth/controller/ll_sw/isoal.c [hunk 4]
```diff
@@ -1088,6 +1114,50 @@ isoal_status_t isoal_rx_pdu_recombine(isoal_sink_handle_t sink_hdl,
 #endif /* CONFIG_BT_CTLR_SYNC_ISO || CONFIG_BT_CTLR_CONN_ISO */
 
 #if defined(CONFIG_BT_CTLR_ADV_ISO) || defined(CONFIG_BT_CTLR_CONN_ISO)
+/**
+ * @breif Check if a time difference calculation is valid and return the difference.
```
Comment (Thalley) [inline subsys/bluetooth/controller/ll_sw/isoal.c:1118]: `@brief`

File: subsys/bluetooth/controller/ll_sw/isoal.c [hunk 5]
```diff
@@ -1088,6 +1114,50 @@ isoal_status_t isoal_rx_pdu_recombine(isoal_sink_handle_t sink_hdl,
...
+ * @param  time_before Subtrahend
+ * @param  time_after  Minuend
+ * @param  result      Difference if valid
+ * @return             Validity
```
Comment (Thalley) [inline subsys/bluetooth/controller/ll_sw/isoal.c:1122]: Please provide information on when a difference is valid (as far as I can tell, if `time_before` is after `time_after`, then it is invalid)

File: subsys/bluetooth/controller/ll_sw/isoal.c [hunk 6]
```diff
@@ -1814,18 +1885,27 @@ static isoal_status_t isoal_tx_framed_produce(struct isoal_source *source,
...
-		LL_ASSERT(actual_grp_ref_point > tx_sdu->time_stamp);
-		time_offset = actual_grp_ref_point - tx_sdu->time_stamp;
+		time_diff_valid = isoal_get_time_diff(tx_sdu->time_stamp,
+					actual_grp_ref_point, &time_diff);
+		LL_ASSERT(time_diff_valid);
+		LL_ASSERT(time_diff > 0);
+		/* Time difference should be less than the maximum possible
+		 * time-offset of 24-bits.
+		 */
+		LL_ASSERT(time_diff <= 0x00FFFFFF);
```
Comment (Thalley) [inline subsys/bluetooth/controller/ll_sw/isoal.c:1907]: `should` and `assert` doesn't seem to go hand in hand. Should this have been a warning? Or should it have been a `shall`?

**Review by Thalley — DISMISSED** [review]
A few comments for the ISO time wrapping. Won't review the unit tests

File: subsys/bluetooth/controller/ll_sw/isoal.c (hunk 1 above)
Comment (niag-Oticon) [reply subsys/bluetooth/controller/ll_sw/isoal.c:51]: Sorry but I don't think that I understood the question. It is ISOAL_TIME_SPAN_FULL / 2.

File: subsys/bluetooth/controller/ll_sw/isoal.c (hunk 6 above)
Comment (niag-Oticon) [reply subsys/bluetooth/controller/ll_sw/isoal.c:1907]: If the time difference exceeds 24-bits there is something very wrong in either the SDU's timestamp or its target event's reference point. So should / shall ---> must and I would say that the assert is valid in this case.

File: subsys/bluetooth/controller/ll_sw/isoal.c (hunk 1 above)
Comment (Thalley) [reply subsys/bluetooth/controller/ll_sw/isoal.c]: Yes, sorry, I misread :|

File: tests/bluetooth/ctrl_isoal/testcase.yaml [hunk 7]
```diff
@@ -1,44 +1,82 @@
...
+      - tx_framed_1_sdu_3_frag_2_pdu
+      - tx_framed_2_sdu_3_frag_4_pdu
+      - tx_framed_2_sdu_3_frag_4_pdu_padding
+      - tx_framed_2_sdu_1_frag_2_pdu_refPoint2
+      - tx_framed_1_sdu_1_frag_1_pdu_refPoint3
+      - tx_framed_1_zero_sdu_1_frag_1_pdu_maxPDU
+      - tx_framed_1_sdu_1_frag_pdu_alloc_err
+      - tx_framed_1_sdu_1_frag_pdu_emit_err
+      - tx_framed_2_sdu_1_frag_pdu_timeout
+      - tx_framed_cis_fra_per_bv07c-
```
Comment (mtpr-ot) [inline tests/bluetooth/ctrl_isoal/testcase.yaml:81]: Why is there a "-" at the end? Doesn't this exclude the test case?

**Review by mtpr-ot — DISMISSED** [review]
I have not reviewed the unit tests in detail, but overall structure, coverage and quality seems very good. I approve on that basis.

File: tests/bluetooth/ctrl_isoal/testcase.yaml (hunk 7 above)
Comment (niag-Oticon) [reply tests/bluetooth/ctrl_isoal/testcase.yaml:81]: Good question! I didn't really consider it as I followed the existing format. 
But Google tells me that it's part of the YAML markup language denoting members of a list. In this case `testcases`.

Comment (mtpr-ot) [reply tests/bluetooth/ctrl_isoal/testcase.yaml:81]: The trailing "-"? I don't think so

Comment (niag-Oticon) [reply tests/bluetooth/ctrl_isoal/testcase.yaml:81]: Nice catch. Didn't notice.

Comment (niag-Oticon) [reply tests/bluetooth/ctrl_isoal/testcase.yaml]: Removed.

**Review by mtpr-ot — DISMISSED** [review]
LGTM

File: subsys/bluetooth/controller/ll_sw/isoal.c [hunk 8]
```diff
@@ -27,6 +29,7 @@
 #define LOG_MODULE_NAME bt_ctlr_isoal
 #include "common/log.h"
 #include "hal/debug.h"
+#include "hal/ticker.h"
```
Comment (cvinayak) [inline subsys/bluetooth/controller/ll_sw/isoal.c:32]: Move this between like this:
```c
#include "util/memq.h"

#include "hal/ticker.h"

#include "pdu.h"
```

**Review by cvinayak — CHANGES_REQUESTED** [review]
Only reviewed isoal.c. Minor nitpick otherwise looks good.

**Review by cvinayak — APPROVED** [review]

**Review by mtpr-ot — APPROVED** [review]
LGTM

**Pull request merged**